design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+guarantees about what information is available from both resources.
+
+
+### Goals
+
+- Introduce identity to the `CertificateRequest` resource which is securely
+    enforced by cert-manager
+- Ensure the identity exposed in `CertificateRequest`s has parity with the
+    upstream Kubernetes `CertificateSigningRequest` resource
+- Enable external tooling or additions to cert-manager that can make use of this
+    identity in a useful way
+
+### Non-Goals
+
+- Discuss how this identity can be used (policy, auditing)
+- Make changes to upstream Kubernetes to implement identiy in cert-manager
+- Although considered below, "passed down" identity is not part of this design
+
+## Proposal
+
+Kubernetes does [not currently support immutable
+fields](https://github.com/kubernetes/enhancements/blob/8b9b994136371f1bc938aabf012f4c45535d684c/keps/sig-api-machinery/20190603-immutable-fields.md)
+for CRDs. The cert-manager webhook will be responsible for populating and
+enforcing identity fields which are present on the spec of `CertificateRequest`
+resources.
+
+The webhook will be responsible for enforcing the following during a CREATE
+operation:
+- No identity fields have been set by the user creating the resource
+- Set identity fields to exactly what is received from the API server in the
+    [UserInfo](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#userinfo-v1-authentication-k8s-io)
+
+The webhook will also responsible for enforcing the following during an UPDATE
+operation:
+- No changes to identity fields are allowed to be made
+
+### API Changes
+
+In order to expose the identity of who created `CertificateRequest` resources,
+these resources must be updated to have parity with the upstream Kubernetes
+certificates API.
+This means that the `CertificateRequest` API type be updated to include the
+following fields in spec, for all API versions:
+
+```go
+	// Username contains the name of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Username string `json:"username,omitempty"`
+	// UID contains the uid of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	UID string `json:"uid,omitempty"`
+	// Groups contains group membership of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +listType=atomic
+	// +optional
+	Groups []string `json:"groups,omitempty"`
+	// Extra contains extra attributes of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Extra map[string][]string `json:"extra,omitempty"`
+```
+
+### Upgrading
+
+Any `CertificateRequest`s that are present in the cluster at the time of
+upgrade, won't and will never have their identity fields populated. This should
+be acceptable.
```

Comment:
Hm - so these fields are being introduced as optional, and a client cannot always _expect_ them to be set? That seems acceptable/expected, but can we call it out very explicitly in the API changes section? We don't want to enforce how an approver behaves in this case (that is up to whoever is implementing it), but it'd be good to call out for those who do implement it